Simplify timestampWrites API

kainino0x opened the pull request:
Summary:

- ```js
  encoder.beginWhicheverPass({
    // ...
    timestampWrites: {
      querySet: myTimestampQuerySet,
      beginningOfPassWriteIndex: 0,
      endOfPassWriteIndex: 1,
    }
  });
  ```
- Both write indices are optional, but at least one is required.
- They may not write to the same index.
- Revisiting this I find myself wondering what the beginning of a "pass write index" is, but this is probably silly.

Fixes #3808

toji commented on spec/index.bs:
```diff
@@ -10248,21 +10253,16 @@ GPUComputePassEncoder includes GPUBindingCommandsMixin;
-    required GPUSize32 queryIndex;
-    required GPUComputePassTimestampLocation location;
+    GPUSize32 beginningOfPassWriteIndex;
+    GPUSize32 endOfPassWriteIndex;
```
These names are very verbose, especially when viewed in context:

```js
encoder.beginRenderPass({
  colorAttachments: attachments,
  timestampWrites: {
    querySet: timestampQueries,
    beginningOfPassWriteIndex: 0,
    endOfPassWriteIndex: 1,
  }
});
```

Given that they're already part of a pass, and already under a key called `timestampWrites`, what about something like `beginIndex` and `endIndex`?

greggman replied:
I have what is probably a dumb question but ... is there a point to being able to specify a begin and end index separate vs just specifying 1 and having it takes 2 slots (and maybe required it to be a align(offset, 2)?`. Is it common to want to put these 2 stamps at unique places?

Advantages

* less to type?
* less to validate?
* simpler to use?

kainino0x replied:
Theoretically one might use the beginning of one pass and the end of a different, later pass. In practice all of these APIs work for an application though, I'd only maybe worry about abstractions on top of WebGPU being limited by our API.

litherum replied:
The current design makes it possible to only write one of the two indices

toji reviewed (approved):
I'm not overtly concerned about the names, honestly. (Verbosity isn't exactly something we've optimized for thus far.) LGTM!

kainino0x commented on spec/index.bs:
```diff
@@ -10599,24 +10599,19 @@ When executing encoded render pass commands as part of a {{GPUCommandBuffer}}, a
-};
-
-dictionary GPURenderPassTimestampWrite {
+dictionary GPURenderPassTimestampWrites {
```
Could be done through dictionary inheritance, but we will want more locations here eventually so they're not going to stay the same forever. Minor preference from me and Myles to keep them separate. (Could be combined later non-breakingly if we really wanted to.)

kainino0x commented on spec/index.bs:
```diff
@@ -12705,13 +12701,15 @@ Issue: Because timestamp query provides high-resolution GPU timestamp, we need t
+    - |timestampWrites|.`querySet` must be [$valid to use with$] |device|.
+    - |timestampWrites|.`querySet`.{{GPUQuerySet/type}} must be {{GPUQueryType/"timestamp"}}.
+    - |timestampWrites|.`beginningOfPassWriteIndex` must be &lt; |timestampWrite|.`querySet`.{{GPUQuerySet/count}}.
+    - |timestampWrites|.`endOfPassWriteIndex` must be &lt; |timestampWrite|.`querySet`.{{GPUQuerySet/count}}.
```
Forgot about these comments. Resolved both, I think it's good to require one index for user clarity.

kainino0x commented:
Addressed comments and updated OP with summary for tacit resolution queue.

> Revisiting this I find myself wondering what the beginning of a "pass write index" is, but this is probably silly.
